File: codes/rust/51.n-皇后.rs
/*
 * @lc app=leetcode.cn id=51 lang=rust
 *
 * [51] N 皇后
 */

// @lc code=start
impl Solution {
    pub fn solve_n_queens(n: i32) -> Vec<Vec<String>> {
        let mut result: Vec<Vec<String>> = Vec::new();
        for i in 0..n {
            let mut board_row = vec!['.'; n as usize];
            let mut board: Vec<String> = vec![board_row.iter().collect::<String>(); n as usize];
            board_row[i as usize] = 'Q';
            board[0] = board_row.iter().collect::<String>();
            Self::backtrace(&mut board, 1, &mut result);
        }
        result
    }

    fn backtrace(board: &mut Vec<String>, row: usize, res: &mut Vec<Vec<String>>) {
        let n = board.len();
        if row == n {
            res.push(board.clone());
            return;
        }
        for col in 0..n {
            if !Self::is_valid(board, row, col) {
                continue;
            }
            // 选择放置皇后
            board[row].replace_range(col..col + 1, "Q");
            Self::backtrace(board, row + 1, res);
            // 撤销选择
            board[row].replace_range(col..col + 1, ".");
        }
    }

    fn is_valid(board: &Vec<String>, row: usize, col: usize) -> bool {
        // 判断正上方
        for i in 0..row {
            if let Some(queue) = board[i].get(col..col + 1) {
                if queue == "Q" {
                    return false;
                }
            }
        }

        // 判断左上方
        for (i, j) in (0..row).rev().zip((0..col).rev()) {
            if let Some(queue) = board[i].get(j..j + 1) {
                if queue == "Q" {
                    return false;
                }
            }
        }

        // 判断右上方
        for (i, j) in (0..row).rev().zip(col + 1..board.len()) {
            if let Some(queue) = board[i].get(j..j + 1) {
                if queue == "Q" {
                    return false;
                }
            }
        }
        true
    }
}
// @lc code=end

File: codes/golang/51.n皇后.go
/*
 * @lc app=leetcode.cn id=51 lang=golang
 *
 * [51] N皇后
 */

 func solveNQueens(n int) [][]string {
	var result [][]string
	board := make([][]rune,n)
	for i:=0; i < n; i++ {
		board[i] = make([]rune,n)
		for j:=0; j < n; j++ {
			board[i][j] = '.'
		}
	}
	backTrace(board, 0, &result)
	return result
}

func backTrace(board [][]rune, row int, result *[][]string) {

	n := len(board)
	if row == n {
		temp := make([]string,n)
		for i:=0; i < n; i++ {
			for j:=0; j < n; j++ {
				temp[i] += string(board[i][j])
			}
		}
		*result = append(*result, temp)
		return
	}
	for col := 0; col < n; col++ {
		if !isValid(board, row, col) {
			//fmt.Println("校验不通过")
			continue
		}
		board[row][col] = 'Q'
		backTrace(board, row+1, result)
		board[row][col] = '.'
	}
}

func isValid(board [][]rune, row int, col int) bool {

	// 这里的行和列表示棋子的坐标
	n := len(board) // 棋盘大小

	// 检查正上方是否有Q
	for i := 0; i < row; i++ {
		if board[i][col] == 'Q' {
			return false
		}
	}

	// 检查右上方有没有皇后
	for i, j := row-1, col+1; i >= 0 && j < n; i, j = i-1, j+1 {
		if board[i][j] == 'Q' {
			return false
		}
	}

	// 检查左上方有没有皇后
	for i, j := row-1, col-1; i >= 0 && j >= 0; i, j = i-1, j-1 {
		if board[i][j] == 'Q' {
			return false
		}
	}

	//// 检查左边有没有Q字符
	//for j:=0; j < col; j=j+1 {
	//	if board[row][j] == 'Q' {
	//		return false
	//	}
	//}
	return true
}

// @lc code=end

